articles/execute_process_refactor.md

```diff
@@ -0,0 +1,597 @@
+| package        | name of the ROS package the node executable lives in         |
+|node_executable|Name of the executable to find|
+| nodes             | A list of `launch_ros.descriptions.Node` objects which are part of the executable. |
+
+Additional parameters that may be passed, which are handled by `launch.actions.ExecuteProcess`: `cmd`, `name`, `cwd`, `env`, `additional_env`.
+NOTE: To allow ROS to determine the appropriate executable based on the package and executable name, the `cmd` parameter should *NOT* be specified. When `cmd` is not specified, this class will determine the appropriate executable if and only if exactly one of the nodes it contains has the `launch_ros.traits.IsExecutable` trait.
+
+#### Properties
+
+Accessors would be provided for the constructor parameters: `package`, `node_executable`, `nodes`.
+Inherited accessors would also be available: `cmd`, `name`, `cwd`, `env`, and `additional_env`.
+
+#### Methods
+
+| Name          | Description                                                  |
+| ------------- | ------------------------------------------------------------ |
+| apply\_context | Overridden. Calls `prepare()` on each `launch_ros.descriptions.Node`, and determines the appropriate executable if necessary. |
+
+In constructing the command line parameters for execution, if more than one node is defined, all appropriate parameters will be prefixed with the node name as necessary.
+
+#### Events
+
+No events would be handled nor emitted.
+
+## Future Extension
+
+One topic of interest is the ability to launch nodes on remote machines as part of more complex systems.
+While the current refactoring will not attempt to implement this functionality, such functionality could be implemented using sibling classes to those described above.
+
+### launch.actions.ExecuteSshProcess
+
+Extends the `launch.actions.Action` class.
+This class would represent the execution-time aspects of an analog to `launch.actions.ExecuteProcess` which executes the given process via an SSH connection.
+This likely would be a remote machine in virtually all usages. This is *not* a subclass of the proposed `launch.actions.ExecuteLocal` class; certain features provided by that class are not available from standard SSH connections.
+
+It is also conceivable that this class, or one like it, could be constructed to simultaneously execute multiple processes over a single SSH connection.
+If this is attempted, several features such as input/output pipes would likely no longer be feasible.
+
+#### Constructor
+
+| Argument               | Description                                                  |
+| ---------------------- | ------------------------------------------------------------ |
+| process\_description   | The `launch.descriptions.Executable` to execute as a remote process |
+| connection_description | An object defining the SSH connection information, such as host, port, user, etc. |
```

Comment:
@roger-strain hmm, one thought (nothing to be changed): if `connection_description` were to become an interface to a given execution environment, a single `launch.actions.Execute()` could later deal with all forms of execution generically. May be overkill though.

Reply:
@hidmic Wanted to follow up on this, as I was having some thoughts on that topic and wanted to run an idea by you before I start writing it up in the formal document. What would you think about going ahead with `launch.actions.Execute()`, which takes two things: the `launch.descriptions.Executable` to actually run, and *optionally* a `launch.descriptions.ExecutionEnvironment` or subclass thereof, which describes how it should be executed. If you pass nothing, local execution is assumed, and a default `launch.descriptions.LocalExecutionEnvironment` object is created under the hood.

You could inspect the object to get a reference to that default object later, if necessary. If you wanted to set any of the parameters currently assigned to `launch.actions.ExecuteLocal`, you would instead create your own `launch.descriptions.LocalExecutionEnvironment` with the appropriate bits.

(All of the following is forward-thinking and not part of the scope of this refactor):
This then would allow you to also define things like `launch.descriptions.SshExecutionEnvironment`, which would open an SSH session using the various parameters passed to it, or `launch.descriptions.PsExecutionEnvironment` if you have something with a PowerShell flavor. Where this could really shine is if you wanted to do something fancy, like have a single SSH session which could run multiple processes. That could look like a `SharedSshExecutionEnvironment` that you pass to all the relevant `launch.actions.Execute` calls, and then at the end a `SharedSshExecutionEnvironment.ExecuteAll()` method would let you launch one SSH session and run all the things that had been queued up inside it.

I think I'd need to mull a little more about where certain things end up, but it seemed like an interesting approach, and I wanted to capture my thoughts and start the discussion before I forgot.

Reply:
At first sight, all of that makes sense to me. If I may, I'd also suggest a `Group`-like action to easily put multiple actions in the same execution environment (akin to `roslaunch`'s `<machine/>` tag).

Reply:
@hidmic After poking around at this for a bit, I'm not sure we're really gaining anything with that aspect of the refactor. Basically everything that is defined in the currently proposed `launch.actions.ExecuteLocal` would have to be migrated into the new `launch.descriptions.LocalExecutionEnvironment`, and several methods which are currently overrides of `launch.actions.Action` would have to be redefined and reimplemented on a base `launch.descriptions.ExecutionEnvironment`. The new `launch.actions.Execute` class would basically be a nearly-empty shell that reroutes its own methods into calling the same-named methods on the `launch.descriptions.ExecutionEnvironment` class. It kind of feels like just adding an extra layer with little to no real benefit.

If it turns out that there would be benefit later, such as when trying to add multi-machine support, we could address it at that point.

Curious as to your thoughts. Assuming we're okay just staying with the design as it stands, do we need any further feedback from other contributors, or is it time to start implementation?